### ISSUE TITLE
Verifying all ATXs before starting node sync takes a lot of time and CPU

### ISSUE THREAD

**Issue by brusherru** [spacemeshos/go-spacemesh#3757]
## Description
Before Node starts syncing layers it downloads and verifies all activations (ATXs).
It causes me to spend **about 80 minutes syncing ATXs** from scratch on devnet-300 and **consumes about 100% of the CPU** within this process. So this is a bug and we can significantly speed up this process.

My most significant concern is that this process will take much longer on the main net if someone wants to join the network after some time from the beginning.
I did some calculations, but it might be wrong. Fix me then :)

**Network stats at the moment:**
- 10 000 layers x 120 seconds per layer
   it lived for ~13.8 days or about 20 000 minutes
- 60 layers per epoch = epoch №166
- about 75 miners,
- up to ~12500 atxs in total for that period
  _but most likely less, since not all miners joined right from the first epoch_

**What it takes**
- It took about 80 minutes to sync up only ATXs from scratch on my machine (M1),
   ~3.84 seconds per ATXs on average
  _some community members wrote in Discord that it takes 20 minutes and 40 minutes, so I believe it also depends on the machine and also the last synced epoch_

**Some extrapolation**
- if we'll have 400 layers per epoch in the main net,
- then for the same period of the running network (~13.8 days)
- there will be 25 epochs
- let's say that we'll have 1000 miners instead of 75 in the main net,
- then there will be about 25000 ATXs within the same period of time

I assume that the required time for syncing has a pretty linear relation
to the amount of ATXs, since we need to take each one after another.
- In this case, it will take about 3 hours to sync up ATXs for the network
that runs for just a couple of weeks.

The time of syncing layers is not included in the calculations at all.
So expect that it will take some more time.

## Steps to reproduce
1. Just start syncing the node from scratch against devnet-300

## Actual Behavior
It consumes a lot of CPU (~100%) and takes a lot of time to sync up ATXs.

## Expected Behavior
It syncs up much faster and does not consume so much CPU.

## Environment
Please complete the following information:
 - OS: macOS 11.6 Big Sur, MacBook Pro (13-inch, M1, 2020). 16GB RAM
 - Node Version: v0.2.18-beta.0

## Out of scope
Since it takes some time to sync up ATXs probably we also need to add the status (like the last synced epoch) in some GRPC API endpoint, like `NodeStatus` and `NodeStatusStream`.

**Comment by dshulyak** [comment]
i think the plan for mainnet is to have 2000 per epoch, 2 weeks/epoch basically. but number of miners is hopefully will be around 1000 - 10000.

regarding cpu, it is expected that consumption will be high during sync. but i haven't looked at sync profiles. by 100% you mean that it utilizes fully 1 core, and not all of them, right?

**Comment by brusherru** [comment]
> by 100% you mean that it utilizes fully 1 core, and not all of them, right?
Yes, right, only one core.

For 2k layers per epoch and having 1k miners it seems much better (about 5.8 days of continuous work) to sync up ATXs for the one-year age network. But, in my humble opinion, it still doesn't sound good, but it is acceptable to have it for genesis. And then hopefully we can optimize it somehow.
But for the testnet (if we won't use the same config, aka layers per epoch) it still sounds unacceptable.

**Comment by dshulyak** [comment]
> For 2k layers per epoch and having 1k miners it seems much better (about 5.8 days of continuous work) to sync up ATXs for the one-year age network. But, in my humble opinion, it still doesn't sound good, but it is acceptable to have it for genesis. And then hopefully we can optimize it somehow.

definitely needs to be profiled. in a year there will be 26 epochs, 1 000 atxs in each. 26 000 atxs in total. my guess would be that it should be doable to download them in a couple of hours at most.

**Comment by brusherru** [comment]
So we'll have something about 10 minutes per layer? Then we can live with it :)
Meanwhile, we'll have to mimic these properties in the testnet.
I mean that with the current testnet parameters it will be very painful to join testnet late

**Comment by noamnelke** [comment]
What @dshulyak said 👍🏻

Regardless of layer length, epochs will be 2 weeks so the number of epochs per year is indeed 26. We can't predict the number of smeshers / ATXs, but 2k-5k in the first year sounds reasonable to me.

**Comment by dshulyak** [comment]
It doesn't make sense to me that it takes so much time, that code path definitely has to be profiled

**Comment by brusherru** [comment]
Btw, sometimes CPU usage jumps up to 3.5-4 / 8 cores.
Process sample on my M1 if it might help:
[sample-0.2.18-atxs.txt](https://github.com/spacemeshos/go-spacemesh/files/10056221/sample-0.2.18-atxs.txt)

And the log is full of such messages:
```
2022-11-21T13:56:07.038+0200	INFO	6a1f7.atxHandler   	validating nipost	{"node_id": "6a1f7a84dfa6d53428aa48a0a2a3ae62574743911b15fa479f285afd8e5cdd85", "module": "atxHandler", "sessionId": "1bdd20a1-a160-4a18-9bed-90aac448cfa9", "expected_challenge_hash": "0x346f9a0ab496680ba073a9a1cefb8bc68ff112383ada903e47d0ec412620b9f9", "atx_id": "084a71cab9c05cb33ea59eaec1dbe2361a8302d934e66e5800f0d7f2cc07789f", "name": "atxHandler"}
2022-11-21T13:56:07.322+0200	INFO	6a1f7.atxHandler   	processing atx	{"node_id": "6a1f7a84dfa6d53428aa48a0a2a3ae62574743911b15fa479f285afd8e5cdd85", "module": "atxHandler", "sessionId": "1bdd20a1-a160-4a18-9bed-90aac448cfa9", "atx_id": "084a71cab9c05cb33ea59eaec1dbe2361a8302d934e66e5800f0d7f2cc07789f", "epoch_id": 232, "atx_node_id": "a8a016fbc6c968ca48bac97336882efeac85cee12a7563cfc8e04405b1410993", "layer_id": 13920, "name": "atxHandler"}
2022-11-21T13:56:07.323+0200	INFO	6a1f7.atxHandler   	atx is valid	{"node_id": "6a1f7a84dfa6d53428aa48a0a2a3ae62574743911b15fa479f285afd8e5cdd85", "module": "atxHandler", "sessionId": "1bdd20a1-a160-4a18-9bed-90aac448cfa9", "atx_id": "084a71cab9c05cb33ea59eaec1dbe2361a8302d934e66e5800f0d7f2cc07789f", "name": "atxHandler"}
2022-11-21T13:56:07.323+0200	INFO	6a1f7.atxHandler   	finished storing atx in epoch	{"node_id": "6a1f7a84dfa6d53428aa48a0a2a3ae62574743911b15fa479f285afd8e5cdd85", "module": "atxHandler", "sessionId": "1bdd20a1-a160-4a18-9bed-90aac448cfa9", "atx_id": "084a71cab9c05cb33ea59eaec1dbe2361a8302d934e66e5800f0d7f2cc07789f", "epoch_id": 232, "name": "atxHandler"}
2022-11-21T13:56:07.323+0200	INFO	6a1f7.atxHandler   	got new atx	{"node_id": "6a1f7a84dfa6d53428aa48a0a2a3ae62574743911b15fa479f285afd8e5cdd85", "module": "atxHandler", "sessionId": "1bdd20a1-a160-4a18-9bed-90aac448cfa9", "atx_id": "084a71cab9c05cb33ea59eaec1dbe2361a8302d934e66e5800f0d7f2cc07789f", "challenge": "0x346f9a0ab496680ba073a9a1cefb8bc68ff112383ada903e47d0ec412620b9f9", "id": "084a71cab9", "sender_id": "a8a016fbc6c968ca48bac97336882efeac85cee12a7563cfc8e04405b1410993", "prev_atx_id": "3127e190da", "pos_atx_id": "3127e190da", "coinbase": "stest1qqqqqqztuw7tq633g22a7cnfwclsxlvg54slhxgyq6a6k", "pub_layer_id": 13920, "epoch": 232, "num_units": 4, "sequence_number": 7, "size": 3164, "name": "atxHandler"}
```

**Comment by dshulyak** [comment]
likely caused by https://github.com/spacemeshos/go-spacemesh/issues/3770

**Comment by dshulyak** [comment]
It should be verified on the version with that bug fixed.

**Comment by brusherru** [comment]
I can't reproduce it anymore. So seems the issue is solved 👍
